external_search.py

```diff
@@ -458,58 +462,93 @@ def query_works(self, query_string, filter=None, pagination=None,
-        # Convert the Search object into a list of hits.
+        :yield: A sequence of lists, one per item in `queries`,
+            each containing the search results from that
+            (query string, Filter, Pagination) 3-tuple.
+        """
+        # If the works alias is not set, all queries return empty.
         #
+        # TODO: Maybe an unset works_alias should raise
+        # CannotLoadConfiguration in the constructor. Then we wouldn't
+        # have to worry about this.
+        if not self.works_alias:
+            for q in queries:
+                yield []
+
+        # Create a MultiSearch. 
+        multi = MultiSearch(using=self.__client)
+
+        # Give it a Search object for every query definition passed in
+        # as part of `queries`.
+        for (query_string, filter, pagination) in queries:
+            search = self.create_search_doc(
+                query_string, filter=filter, pagination=pagination, debug=debug
+            )
+            function_scores = filter.scoring_functions if filter else None
+            if function_scores:
+                function_score = FunctionScore(
+                    query=dict(match_all=dict()),
+                    functions=function_scores,
+                    score_mode="sum"
+                )
+                search = search.query(function_score)
+            multi = multi.add(search)
+
+        a = time.time()
         # NOTE: This is the code that actually executes the ElasticSearch
         # request.
-        results = [x for x in results]
+        resultset = [x for x in multi.execute()]
 
         if debug:
             b = time.time()
             self.log.debug(
                 "Elasticsearch query %r completed in %.3fsec",
                 query_string, b-a
             )
-            for i, result in enumerate(results):
-                self.log.debug(
-                    '%02d "%s" (%s) work=%s score=%.3f shard=%s',
-                    i, result.sort_title, result.sort_author, result.meta['id'],
-                    result.meta['score'] or 0, result.meta['shard']
-                )
+            for results in resultset:
+                for i, result in enumerate(results):
+                    self.log.debug(
+                        '%02d "%s" (%s) work=%s score=%.3f shard=%s',
+                        i, result.sort_title, result.sort_author, result.meta['id'],
+                        result.meta['score'] or 0, result.meta['shard']
+                    )
 
-        # Tell the Pagination object about this page -- this may help
-        # it set up to generate a link to the next page.
-        pagination.page_loaded(results)
-
-        return results
+        # Process each item in the resultset according to the
+        # Pagination object that was used to create it.
+        for i, results in enumerate(resultset):
+            # Use the Pagination object to slice up the results if
+            # necessary.
+            results = resultset[i]
```

Comment:
Why use `resultset[i]` when the loop has `i, results` above?

Reply:
Good catch.